File: services/api/src/bootstrap/required-secrets.ts
/**
 * Shared Required_Secret bootstrap validator for the API_Service.
 *
 * Runs BEFORE Fastify and the application logger are initialized, so all
 * output goes through `process.stderr.write` and termination is handled by
 * `process.exit(1)`. The functions never throw — they either exit non-zero
 * (production) or return after logging a warning (development/test).
 */

export type SecretSpec = { name: string; minLength?: number };
export type IntegerSpec = { name: string; min?: number };

/**
 * Canonical list of Required_Secrets for the API_Service. Kept in sync with
 * the design document (Wave B / R1) and `.env.example`.
 */
export const API_REQUIRED_SECRETS: SecretSpec[] = [
  { name: "DATABASE_URL" },
  { name: "REDIS_URL" },
  { name: "JWT_ACCESS_SECRET", minLength: 32 },
  { name: "JWT_REFRESH_SECRET", minLength: 32 },
  { name: "INTERNAL_SERVICE_TOKEN", minLength: 16 },
  { name: "CORS_ORIGINS" },
];

function isProduction(env: NodeJS.ProcessEnv): boolean {
  return env.NODE_ENV === "production";
}

function reportAndExit(env: NodeJS.ProcessEnv, label: string, offenders: string[]): void {
  if (offenders.length === 0) return;
  const names = offenders.join(", ");
  if (isProduction(env)) {
    process.stderr.write(`[bootstrap] Missing or invalid ${label}: ${names}\n`);
    process.exit(1);
    return;
  }
  process.stderr.write(
    `[bootstrap] WARN: Missing or invalid ${label}: ${names} (dev mode, continuing)\n`,
  );
}

/**
 * Verify that every Required_Secret in `specs` is present, non-empty (after
 * trimming whitespace) and — when `minLength` is set — at least that long.
 *
 * Production failure: write a single line listing every offender to stderr
 * and call `process.exit(1)`. Dev/test failure: write a warning line and
 * return.
 */
export function assertRequiredSecrets(specs: SecretSpec[], env: NodeJS.ProcessEnv): void {
  const offenders: string[] = [];
  for (const spec of specs) {
    const raw = env[spec.name];
    if (raw === undefined || raw === null || raw.trim().length === 0) {
      offenders.push(spec.name);
      continue;
    }
    if (spec.minLength !== undefined && raw.length < spec.minLength) {
      offenders.push(`${spec.name} (min length ${spec.minLength})`);
    }
  }
  reportAndExit(env, "Required_Secret(s)", offenders);
}

/**
 * Verify that every variable in `specs` parses to a positive integer
 * (no negatives, no zero, no NaN, no decimals). Production failure: exit
 * non-zero. Dev/test failure: warn and return.
 */
export function assertPositiveIntegers(specs: IntegerSpec[], env: NodeJS.ProcessEnv): void {
  const offenders: string[] = [];
  for (const spec of specs) {
    const raw = env[spec.name];
    if (raw === undefined || raw === null || raw.trim().length === 0) {
      offenders.push(spec.name);
      continue;
    }
    const trimmed = raw.trim();
    // Reject anything that is not a plain decimal-integer literal (no dots,
    // no exponents, no leading sign except an explicit positive int).
    if (!/^\d+$/.test(trimmed)) {
      offenders.push(`${spec.name} (not a positive integer: ${raw})`);
      continue;
    }
    const parsed = Number.parseInt(trimmed, 10);
    if (!Number.isFinite(parsed) || parsed <= 0) {
      offenders.push(`${spec.name} (must be > 0)`);
      continue;
    }
    if (spec.min !== undefined && parsed < spec.min) {
      offenders.push(`${spec.name} (must be >= ${spec.min})`);
    }
  }
  reportAndExit(env, "positive integer Required_Secret(s)", offenders);
}

File: services/ai/bootstrap/required_secrets.py
"""
AI service bootstrap-time validators.

This module is intentionally side-effect-free on import: the
``assert_*`` functions only act (write to ``sys.stderr`` and/or call
``sys.exit(1)``) when explicitly invoked. It does not import FastAPI,
the project's routers, or the standard ``logging`` module — at the
point this code runs, ``logging`` is typically not yet configured.

Production mode is detected from either ``NODE_ENV=production`` or
``ENV=production`` since several of our deployment targets use one or
the other.

Contract used by ``services/ai/main.py`` (task B.2):
    - In production: collect all offending variables, write a single
      line to ``sys.stderr`` listing every offender, and call
      ``sys.exit(1)``.
    - In dev/test: write a single ``WARN`` line and return — never
      raise — so local workflows are not blocked.
"""

from __future__ import annotations

from dataclasses import dataclass
import sys
from typing import Iterable, List, Mapping, Optional


@dataclass(frozen=True)
class SecretSpec:
    """A required secret env var spec.

    ``min_length`` is optional; when present, values shorter than this
    are considered invalid in addition to the standard unset / empty /
    whitespace-only checks.
    """

    name: str
    min_length: Optional[int] = None


@dataclass(frozen=True)
class IntegerSpec:
    """A required positive-integer env var spec.

    ``min`` is optional; when omitted, the validator enforces a
    strictly-positive integer (>= 1). Non-numeric, NaN, decimal, and
    negative values are always rejected.
    """

    name: str
    min: Optional[int] = None


def _is_production(env: Mapping[str, str]) -> bool:
    """Treat both ``NODE_ENV=production`` and ``ENV=production`` as production."""
    return env.get("NODE_ENV") == "production" or env.get("ENV") == "production"


def _emit_failure(prefix: str, offenders: List[str], env: Mapping[str, str]) -> None:
    """Write the single-line stderr message and exit(1) in prod, warn-and-return in dev."""
    names = ", ".join(offenders)
    if _is_production(env):
        sys.stderr.write(
            f"[bootstrap] Missing or invalid {prefix}: {names}\n"
        )
        sys.exit(1)
    else:
        sys.stderr.write(
            f"[bootstrap] WARN: Missing or invalid {prefix}: {names} (dev mode, continuing)\n"
        )


def assert_required_secrets(
    specs: Iterable[SecretSpec],
    env: Mapping[str, str],
) -> None:
    """Validate each spec against ``env``.

    A spec is considered offending when the env value is ``None``,
    empty, whitespace-only, or — when ``min_length`` is set — shorter
    than ``min_length``.

    In production: writes a single line to ``sys.stderr`` listing every
    offender and calls ``sys.exit(1)``. In dev/test: writes a warning
    line and returns. Never raises.
    """
    offenders: List[str] = []
    for spec in specs:
        value = env.get(spec.name)
        if value is None:
            offenders.append(spec.name)
            continue
        stripped = value.strip()
        if stripped == "":
            offenders.append(spec.name)
            continue
        if spec.min_length is not None and len(value) < spec.min_length:
            offenders.append(spec.name)
            continue

    if not offenders:
        return

    _emit_failure("Required_Secret(s)", offenders, env)


def assert_positive_integers(
    specs: Iterable[IntegerSpec],
    env: Mapping[str, str],
) -> None:
    """Validate that each spec resolves to a positive integer in ``env``.

    Rejects unset, empty/whitespace, non-numeric, negative, zero, NaN,
    and decimal/float values. ``int(value)`` is intentionally NOT used
    on the raw string because Python's ``int("1.5")`` raises but
    ``int(float("1.5"))`` truncates — we want to reject ``"1.5"`` as
    invalid input rather than silently coerce. We therefore parse via
    a strict integer regex.

    In production: writes a single line to ``sys.stderr`` listing every
    offender and calls ``sys.exit(1)``. In dev/test: writes a warning
    line and returns. Never raises.
    """
    offenders: List[str] = []
    for spec in specs:
        raw = env.get(spec.name)
        if raw is None:
            offenders.append(spec.name)
            continue
        stripped = raw.strip()
        if stripped == "":
            offenders.append(spec.name)
            continue

        # Strict integer parse: optional leading '+' then one or more digits.
        # This rejects "1.5", "nan", "inf", "1e3", " 1 2 ", etc.
        if not (stripped.lstrip("+").isdigit()):
            offenders.append(spec.name)
            continue

        try:
            value = int(stripped, 10)
        except ValueError:
            offenders.append(spec.name)
            continue

        threshold = spec.min if spec.min is not None else 1
        if value < threshold:
            offenders.append(spec.name)
            continue

    if not offenders:
        return

    _emit_failure("positive integer env var(s)", offenders, env)


# ---------------------------------------------------------------------------
# Spec lists for AI_Service. ``main.py`` imports and uses these directly.
# ---------------------------------------------------------------------------

AI_REQUIRED_SECRETS: List[SecretSpec] = [
    SecretSpec("OPENROUTER_API_KEY"),
    SecretSpec("INTERNAL_SERVICE_TOKEN", min_length=16),
]

AI_RATE_LIMIT_THRESHOLDS: List[IntegerSpec] = [
    IntegerSpec("AI_FREE_MESSAGES_PER_DAY"),
    IntegerSpec("AI_BASIC_MESSAGES_PER_DAY"),
    IntegerSpec("AI_PREMIUM_MESSAGES_PER_DAY"),
]
